### ISSUE TITLE
💅 `noLeakedRender` many false positives

### ISSUE THREAD

**Issue by jakeleventhal** [biomejs/biome#8288]
### Environment information

```bash
CLI:
  Version:                      2.3.8
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v22.21.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.11.1

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementInteractions
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExcessiveNestedTestSuites
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noImportantStyles
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringConcat
    complexity/noUselessStringRaw
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalDirnameFilename
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNestedComponentDefinitions
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noReactPropAssignments
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUndeclaredDependencies
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useGraphqlNamedOperations
    correctness/useIsNan
    correctness/useParseIntRadix
    correctness/useSingleJsDocAsterisk
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    nursery/noDeprecatedImports
    nursery/noDuplicateDependencies
    nursery/noDuplicatedSpreadProps
    nursery/noEmptySource
    nursery/noEqualsToNull
    nursery/noFloatingPromises
    nursery/noForIn
    nursery/noImportCycles
    nursery/noLeakedRender
    nursery/noMisusedPromises
    nursery/noNextAsyncClientComponent
    nursery/noParametersOnlyUsedInRecursion
    nursery/noProto
    nursery/noReactForwardRef
    nursery/noSyncScripts
    nursery/noUnknownAttribute
    nursery/noUnusedExpressions
    nursery/noUselessCatchBinding
    nursery/noUselessUndefined
    nursery/useConsistentArrowReturn
    nursery/useConsistentGraphqlDescriptions
    nursery/useExhaustiveSwitchCases
    nursery/useFind
    nursery/useSortedClasses
    nursery/useSpread
    nursery/useUniqueGraphqlOperationName
    performance/noDelete
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noGlobalEval
    style/noCommonJs
    style/noDescendingSpecificity
    style/noInferrableTypes
    style/noNestedTernary
    style/noParameterAssign
    style/noParameterProperties
    style/noRestrictedImports
    style/noSubstr
    style/noUnusedTemplateLiteral
    style/noUselessElse
    style/noYodaExpression
    style/useArrayLiterals
    style/useAsConstAssertion
    style/useAtIndex
    style/useBlockStatements
    style/useCollapsedElseIf
    style/useCollapsedIf
    style/useConsistentArrayType
    style/useConsistentCurlyBraces
    style/useConsistentMemberAccessibility
    style/useConsistentObjectDefinitions
    style/useConsistentTypeDefinitions
    style/useConst
    style/useDefaultParameterLast
    style/useDefaultSwitchClause
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useForOf
    style/useFragmentSyntax
    style/useGraphqlNamingConvention
    style/useGroupedAccessorPairs
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodeAssertStrict
    style/useNodejsImportProtocol
    style/useNumberNamespace
    style/useObjectSpread
    style/useReactFunctionComponents
    style/useReadonlyClassProperties
    style/useSelfClosingElements
    style/useShorthandAssign
    style/useShorthandFunctionType
    style/useSingleVarDeclarator
    style/useTemplate
    style/useThrowNewError
    style/useThrowOnlyError
    style/useTrimStartEnd
    suspicious/noAlert
    suspicious/noApproximativeNumericConstant
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noBitwiseOperators
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConsole
    suspicious/noConstEnum
    suspicious/noConstantBinaryExpressions
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noDuplicateTestHooks
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSkippedTests
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noVar
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useErrorMessage
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useGuardForIn
    suspicious/useIsArray
    suspicious/useNamespaceKeyword
    suspicious/useNumberToFixedDigitsArgument

Workspace:
  Open Documents:               0
```

### Rule name

`noLeakedRender`

### Playground link

https://biomejs.dev/playground/?lineWidth=120&indentStyle=space&lintRules=noLeakedRender&code=ZgB1AG4AYwB0AGkAbwBuACAAZwBlAHQAVgBhAGwAdQBlACgAKQA6ACAAYgBvAG8AbABlAGEAbgAgAHsACgAgACAAcgBlAHQAdQByAG4AIABNAGEAdABoAC4AcgBhAG4AZABvAG0AKAApACAAPgAgADAALgA1ADsACgB9AAoACgBlAHgAcABvAHIAdAAgAGQAZQBmAGEAdQBsAHQAIABmAHUAbgBjAHQAaQBvAG4AIABNAHkAQwBvAG0AcABvAG4AZQBuAHQAKAApACAAewAKACAAIAByAGUAdAB1AHIAbgAgACgACgAgACAAIAAgADwAYgB1AHQAdABvAG4AIABvAG4AQwBsAGkAYwBrAD0AewBnAGUAdABWAGEAbAB1AGUAKAApACAAPwAgAGcAZQB0AFYAYQBsAHUAZQAgADoAIAB1AG4AZABlAGYAaQBuAGUAZAB9ACAALwA%2BAAoAIAAgACkACgB9AA%3D%3D

### Expected result

No error this is a valid ternary that has no risk of leaked rendering.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by ematipico** [comment]
How does the eslint plugin behave? 

Also, are there other examples you missed? The title says "many"... :D

**Comment by ematipico** [comment]
cc @dibashthapa

**Comment by jakeleventhal** [comment]
@ematipico many because my entire codebase exploded when i added this rule lol. probably just one root cause

**Comment by dibashthapa** [comment]
I will add these test cases tomorrow

**Comment by ceolinwill** [comment]
> my entire codebase exploded when i added this rule

I think the issue might be that the rule doesn’t treat TypeScript booleans as booleans. For example:

```ts
// `pending` is a boolean
const { pending } = useFormStatus();

{pending && <Spinner />}
```

This gets flagged, but it only passes if I do:

```ts
{Boolean(pending) && <Spinner />}
```

Which doesn’t change anything because `pending` is already typed as a boolean.

I’m not sure whether this is expected or a bug, though.

**Comment by dibashthapa** [comment]
Yeah you are right. The original rule also suggest it to not use on typed codebase. 

However for this specific code , eslint doesn't show any errors because it is still valid ternary expression in js. 

I think it is happening because I am not handling the alternate values correctly. This should be fixed after making some tiny changes and adding new test cases

**Comment by dibashthapa** [comment]
@jakeleventhal do you have other examples where you are having this issue ?

**Comment by dibashthapa** [comment]
Currently, 

This also triggers errors 

```jsx
const MyComponent4 = () => {
	return <Something checked={v ? false : isChecked} />;
};

const MyComponent5 = () => {
	return <Something checked={cond && isIndeterminate ? false : isChecked} />;
};
```

These are also false positive errors , which makes me think if we should completely ignore checking on ternary expressions

Because i can't think of any examples that could leak values in ternary expressions , and also because this is predictable , and makes more  unlike conditional expressions with `&&`, where the most issues occur

```jsx
const MyComponent = () => {
	return <Something checked={v ? false : null} />;
};

const MyComponent = () => {
	return <Something checked={v ? false : 0} />;
};

const MyComponent = () => {
	return <Something checked={v ? 0 : undefined} />;
};
```

@ematipico what do you think of this ?

**Comment by jakeleventhal** [comment]
@dibashthapa 

[all-errors.txt](https://github.com/user-attachments/files/23824940/all-errors.txt)

**Comment by dibashthapa** [comment]
Thanks @jakeleventhal

**Comment by dyc3** [comment]
It seems like there are multiple bugs here, but going back to the playground:

```tsx
export default function MyComponent() {
  return (
    <button onClick={getValue() ? getValue : undefined} />
  )
}
```
This a false positive because the expression `{getValue() ? getValue : undefined}` is not rendered at all, it is a prop. As in, its not a direct child of a `JsxElementList`.

**Comment by dibashthapa** [comment]
Yeah, I am going to ignore JsxAtttribute
